pennylane/beta/tapes/qnode.py

```diff
@@ -0,0 +1,541 @@
+    """
+
+    # pylint:disable=too-many-instance-attributes
+
+    def __init__(self, func, device, interface="autograd", diff_method="best", **diff_options):
+
+        if interface is not None and interface not in self.INTERFACE_MAP:
+            raise QuantumFunctionError(
+                f"Unknown interface {interface}. Interface must be "
+                f"one of {self.INTERFACE_MAP.values()}."
+            )
+
+        if not isinstance(device, Device):
+            raise QuantumFunctionError("Invalid device. Device must be a valid PennyLane device.")
+
+        self.func = func
+        self.device = device
+        self.qtape = None
+
+        self._tape, self.interface, self.diff_method = self._get_tape(
+            device, interface, diff_method
+        )
+        self.diff_options = diff_options or {}
+        self.diff_options["method"] = self.diff_method
+
+        self.dtype = np.float64
+        self.max_expansion = 2
+
+    @staticmethod
+    def _get_tape(device, interface, diff_method="best"):
+        """Determine the best QuantumTape, differentiation method, and interface
+        for a requested device, interface, and diff method.
+
+        Args:
+            device (.Device): PennyLane device
+            interface (str): name of the requested interface
+            diff_method (str): The requested method of differentiation. One of
+                ``"best"``, ``"backprop"``, ``"reversible"``, ``"device"``,
+                ``"parameter-shift"``, or ``"finite-diff"``.
+
+        Returns:
+            tuple[.QuantumTape, str, str]: tuple containing the compatible
+            QuantumTape, the interface to apply, and the method argument
+            to pass to the ``QuantumTape.jacobian`` method.
+        """
+
+        if diff_method == "best":
+            return QNode._get_best_method(device, interface)
+
+        if diff_method == "backprop":
+            return QNode._validate_backprop_method(device, interface)
+
+        if diff_method == "device":
+            return QNode._validate_device_method(device, interface)
+
+        if diff_method == "parameter-shift":
+            return QNode._get_parameter_shift_method(device, interface)
+
+        if diff_method == "finite-diff":
+            return QuantumTape, interface, "numeric"
+
+        raise QuantumFunctionError(
+            f"Differentiation method {diff_method} not recognized. Allowed "
+            "options are ('best', 'parameter-shift', 'backprop', 'finite-diff', 'device', 'reversible')."
+        )
+
+    @staticmethod
+    def _get_best_method(device, interface):
+        """Returns the 'best' QuantumTape and differentiation method
+        for a particular device and interface combination.
+
+        This method attempts to determine support for differentiation
+        methods using the following order:
+
+        * ``"backprop"``
+        * ``"device"``
+        * ``"parameter-shift"``
+        * ``"reversible"``

```

Comment:
Are we waiting on reversible for a follow up PR?

Reply:
yep!